### PR TITLE
Contribute new pack for Cisco ISE

### PR DESCRIPTION
**Pull request by namachieli**
A Pack that uses requests and primarily yaml actions to interact with the Cisco ISE API.

### CODE REVIEW

**Comment by namachieli** [comment]
Pending review and feedback. Otherwise pack is ready to merge.

**Comment by namachieli** [comment]
@blag @nmaludy Please review and provide feedback. Thank you!

File: cisco_ise/config.schema.yaml
Comment (cognifloyd) [inline cisco_ise/config.schema.yaml:8]: I think it would be better to put these properties under additionalProperties.
https://github.com/StackStorm/st2/pull/5225 fixed the ability to retrieve encrypted secrets from the datastore under `additionalProperties`. I don't think anything checks for `patternProperties`.

Modifying your example, I don't think this will work:

```yaml
connections:
  ise_server:
    api_user: username
    api_pass: "{{ st2kv.system.ise_server_api_pass }}"
    hostname: ise-pan.example.com
```

Also, what happens if someone does this? Technically I think passes the schema (possibly needs one entry in connections):
```yaml
connections: {}
connection_name:
  api_user: username
  api_pass: password
  hostname: ise-pan.example.com
```

File: cisco_ise/actions/lib/base.py
Comment (cognifloyd) [inline cisco_ise/actions/lib/base.py:275]: What about setting `self.action_kwargs = None` in `__init__()` and then:

```suggestion
        if self.action_kwargs is None:
            raise NameError(
                '_prepare_kwargs() is required to be ran prior to _setup_connection()')
```

File: cisco_ise/actions/lib/base.py
Comment (cognifloyd) [inline cisco_ise/actions/lib/base.py:281]:
```suggestion
            connection = self.retrieve_config_connection(connection_name=self.connection_name)
```

File: cisco_ise/actions/lib/base.py
Comment (cognifloyd) [inline cisco_ise/actions/lib/base.py:292]:
```suggestion
            # If we haven't found the key yet, check if it's required
```

File: cisco_ise/actions/lib/base.py
Comment (cognifloyd) [inline cisco_ise/actions/lib/base.py:382]: `if self.connection is None:` like my comment in _setup_connection.

File: cisco_ise/actions/lib/base.py
Comment (cognifloyd) [inline cisco_ise/actions/lib/base.py:398]: If protocol is `http`, does this need a different port?

File: cisco_ise/actions/lib/base.py
Comment (cognifloyd) [inline cisco_ise/actions/lib/base.py:426]: `if self.action_kwargs is None`

File: cisco_ise/actions/lib/base.py
Comment (cognifloyd) [inline cisco_ise/actions/lib/base.py:95]:
```suggestion
            payload_format = "json" if self.payload_format is None else self.payload_format
```

File: cisco_ise/actions/lib/base.py
Comment (cognifloyd) [inline cisco_ise/actions/lib/base.py:152]:
```suggestion
                    return self.format_action_result(result,
                                                     response_code,
                                                     succeeded,
                                                     response_msg)
```
`break` is not needed after `return`.

File: cisco_ise/actions/lib/base.py
Comment (cognifloyd) [inline cisco_ise/actions/lib/base.py:119]: You could do `while not last:` here and then move the `return` block outside of the while loop.

File: cisco_ise/actions/lib/base.py
Comment (cognifloyd) [inline cisco_ise/actions/lib/base.py:252]:
```suggestion
            return {}
```

**Review by cognifloyd — CHANGES_REQUESTED** [review]
I don't know ISE, so I can't evaluate any of the API usage.

However, I think using a common python entry point for all of these actions is a pretty sweet idea. I like how you've minimized things.

Do you have a script to generate all off the action metadata files from some kind of openapi spec? Or did you make all of them by hand?

I have several comments on python coding style. These are my gut reactions - I hope the feedback is helpful. Also, could you run `black` on your python code? We've adopted black in core StackStorm, so that makes it a bit easier (more consistent) to review.

**Comment by iprice-salesforce** [comment]
> Do you have a script to generate all off the action metadata files from some kind of openapi spec? Or did you make all of them by hand?

All by hand. Unfortunately the ISE API itself a bit inconsistent, and the documentation is even more so. It is rife with typos, incomplete data, missing structure, and some bit that are just plain wrong.

> Also, could you run black on your python code?

I am not familiar, but I will investigate. Thank you for the heads up!

> I think using a common python entry point for all of these actions is a pretty sweet idea. I like how you've minimized things.

Thank you! 😄 


> These are my gut reactions - I hope the feedback is helpful.

Feedback is always appreciated! I will review and reply to each, thank you for taking the time to do so.

File: cisco_ise/actions/lib/base.py
Comment (iprice-salesforce) [reply cisco_ise/actions/lib/base.py:275]: Agreed. This is more explicit and cleaner.

File: cisco_ise/actions/lib/base.py
Comment (namachieli) [reply cisco_ise/actions/lib/base.py:281]: Noted and fixing this typo globally.

File: cisco_ise/actions/lib/base.py
Comment (namachieli) [reply cisco_ise/actions/lib/base.py:398]: good catch. Fixing.

File: cisco_ise/actions/lib/base.py
Comment (namachieli) [reply cisco_ise/actions/lib/base.py:152]: good catch, cruft left over from a loop rewrite.

File: cisco_ise/actions/lib/base.py
Comment (namachieli) [reply cisco_ise/actions/lib/base.py:119]: I agree, that would be more explicit and easier to understand.

File: cisco_ise/config.schema.yaml
Comment (namachieli) [reply cisco_ise/config.schema.yaml:8]: Good feedback, I will give this a review and make some adjustments.

File: cisco_ise/actions/lib/base.py
Comment (cognifloyd) [inline cisco_ise/actions/lib/base.py:40]:
```suggestion
        self.connection = None
        self.payload_format = None
```
And the other properties added in `_prepare_kwargs`
Generally I like to see the attributes in `__init__` or as class-level attributes before using/modifying them in methods like `_prepare_kwargs`. this looks like a decent explanation: https://stackoverflow.com/a/19292653

**Review by cognifloyd — COMMENTED** [review]
One more minor nit, and then I think this is good to go. :)
Note: You'll still need to wait for someone on the TSC to migrate this over to its own repo.

File: cisco_ise/actions/lib/base.py
Comment (namachieli) [reply cisco_ise/actions/lib/base.py:40]: If I understand...

Methods used to breakup init is fine. Changing attributes in one of these init setup methods is fine. Creating the attribute in the init sub method... not desirable. 

We want to always define all attributes at the head of init, even if we immediately change it in the init logic.

If I have that correct then I am 100% on board and that is a superior way of handling it.

Comment (namachieli) [reply cisco_ise/actions/lib/base.py:43]: I have added all attributes to the main `__init__()` method as `None` and added a comment. Good suggestion, thank you!

File: cisco_ise/config.schema.yaml
Comment (namachieli) [reply cisco_ise/config.schema.yaml:8]: I'm having trouble with how this should be done to make it better, would you be willing to offer a suggestion?

Comment (cognifloyd) [reply cisco_ise/config.schema.yaml:8]: Here's how I did it for the vault pack (not upstreamed yet): https://github.com/copartit/stackstorm-vault/blob/dev/config.schema.yaml#L20

**Review by cognifloyd — APPROVED** [review]
I think this has incubated well enough. I can't actually test the pack, but it looks good to me.

File: cisco_ise/config.schema.yaml
Comment (namachieli) [reply cisco_ise/config.schema.yaml:8]: I think for now I would prefer to leave it as is. I know it at least works as intended and I'm not sure I'll have time to modify and test it properly. Hopefully that is ok.

**Comment by namachieli** [comment]
@blag @nmaludy @cognifloyd 

Anything else needed before this is approved?

**Comment by cognifloyd** [comment]
@namachieli A lot of the maintainers have been very busy with work and other things, so merging stuff can take a really long time. For example, given blag's work commitments he just announced that he's stepping down as a maintainer. And nmaludy switched jobs not too long ago, so he's been swamped with lots to do. Sorry about the delays getting this in!

I can't officially approve this or create the git repo it needs to be in. Hopefully, someone will have time to do it sooner rather than later.

Thank you for your patience, and, again, sorry this is taking awhile.

**Comment by namachieli** [comment]
Understood, thank you for replying. It's a tough situation for all.

**Comment by blag** [comment]
Just a note: the end goal of PRs here isn't actually to get merged.

Somebody with enough access needs to either run this script:
https://github.com/StackStorm-Exchange/ci/blob/master/utils/exchange-bootstrap.sh

Or follow those steps to get this bootstrapped into StackStorm Exchange.

Since I'm leaving, I think somebody else should take up the mantle here. I'm happy to lend a hand if they run into trouble, but I'd like to pass this knowledge on to somebody else to maintain that ability in the TSC.

**Comment by armab** [comment]
Closing as https://github.com/stackStorm-exchange/stackstorm-cisco_ise pack is available in the StackStorm Index: https://exchange.stackstorm.org/#cisco_ise

**Pull request closed without merging**
